test/test_in_toto_record.py

```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env python
+
+"""
+<Program Name>
+  test_in_toto_record.py
+
+<Author>
+  Lukas Puehringer <lukas.puehringer@nyu.edu>
+
+<Started>
+  Nov 28, 2016
+
+<Copyright>
+  See LICENSE for licensing information.
+
+<Purpose>
+  Test in_toto_record command line tool.
+
+"""
+
+import os
+import sys
+import unittest
+import logging
+import argparse
+import shutil
+import tempfile
+from mock import patch
+
+from toto.util import generate_and_write_rsa_keypair
+from toto.models.link import Link
+from toto.in_toto_record import main as in_toto_record_main
+from toto.in_toto_record import in_toto_record_start, in_toto_record_stop
+
+WORKING_DIR = os.getcwd()
+
+# Suppress all the user feedback that we print using a base logger
+logging.getLogger().setLevel(logging.CRITICAL)
+
+class TestInTotoRecordMain(unittest.TestCase):
+  """Test in_toto_record's main() - requires sys.argv patching. """
+
+  @classmethod
+  def setUpClass(self):
+    """Create and change into temporary directory,
+    generate key pair, dummy artifact and base arguments. """
+    self.test_dir = tempfile.mkdtemp()
+    os.chdir(self.test_dir)
+
+    self.test_key = "test_key"
+    generate_and_write_rsa_keypair(self.test_key)
+
+    self.test_artifact = "test_artifact"
+    open(self.test_artifact, "w").close()
+
+    self.args = [
+        "in_toto_record.py",
+        "--step-name",
+        "test-step",
+        "--key",
+        self.test_key
+    ]
+
+  @classmethod
+  def tearDownClass(self):
+    """Change back to initial working dir and remove temp test directory. """
+    os.chdir(WORKING_DIR)
+    shutil.rmtree(self.test_dir)
+
+  def test_command_start_stop_required_args(self):
+    """Test CLI command record start/stop with required arguments. """
+    with patch.object(sys, 'argv',  self.args + ["start"]):
+      in_toto_record_main()
+
+    with patch.object(sys, 'argv', self.args + ["stop"]):
+      in_toto_record_main()
+
+  def test_command_start_stop_optional_args(self):
+    """Test CLI command record start/stop with optional arguments. """
+    with patch.object(sys, 'argv', self.args + ["start", "--materials",
+        self.test_artifact]):
+      in_toto_record_main()
+
+    with patch.object(sys, 'argv', self.args + ["stop", "--products",
+        self.test_artifact]):
+      in_toto_record_main()
```

Comment:
So this is what I don't understand, this is a script that has a `main` function and only runs that `main` function. Why do we declare `main` in the first place? is it going to be called from somewhere else (e.g., `import x; x.main()`)? if we do this, then would a caller need to mock sys also? Wouldn't it be more explicit to have main's functionality inside the if? My feeling is that this hinders functional separation, and that's why we are needing these mocks.